File: R/rnormTest.R

################################################################################
################## Functions for testing random generators #####################
################################################################################

boxMuller <- function()
{
  U <- runif(2)
  root <- sqrt( -2 * log(U[1]) )
  N1 <- root * cos(2 * pi * U[2])
  N2 <- root * sin(2 * pi * U[2])
  c(N1, N2)
}

polarRejection <- function()
{
  U <- runif(2)
  V <- 2*U - 1
  W <- sum(V^2)
  if (W > 1)
    return (polarRejection())
  root <- sqrt( -2 * log(W) / W )
  N1 <- root * V[1]
  N2 <- root * V[2]
  list(fst = N1, snd = N2)
}

testNormalGenerators <- function(n)
{
  bm <- system.time( for (i in 1:n) boxMuller() )
  pr <- system.time( for (i in 1:n) polarRejection() )
  print(paste("Box-Muller:", round(bm[1], 2), "seconds."))
  print(paste("Polar rejection method:", round(pr[1], 2), "seconds."))
}

################################################################################
# inne podejscie

# boxMuller <- function(n)
# {
#   U1 <- runif(n)
#   U2 <- runif(n)
#   roots <- sqrt( -2 * log(U1) )
#   N1 <- roots * cos(2 * pi * U2)
#   N2 <- roots * sin(2 * pi * U2)
#   c(N1, N2)
# }
# 
# polarRejection <- function(n)
# {
#   V1 <- 2*runif(1.5*n) - 1
#   V2 <- 2*runif(1.5*n) - 1
#   W <- V1^2 + V2^2
#   hit <- which(W <= 1)
#   V1 <- (V1[hit])[1:n]
#   V2 <- (V2[hit])[1:n]
#   W <- (W[hit])[1:n]
#   root <- sqrt( -2 * log(W) / W )
#   N1 <- root * V1
#   N2 <- root * V2
#   c(N1, N2)
# }
# 
# testNormalGenerators <- function(n)
# {
#   bm <- system.time( boxMuller(n) )
#   pr <- system.time( polarRejection(n) )
#   print(paste("Box-Muller:", round(bm[1], 2), "seconds."))
#   print(paste("Polar rejection method:", round(pr[1], 2), "seconds."))
# }

File: NBfiles/src/not_app_related/RNormTest.java

package not_app_related;

import static java.lang.Math.*;
import java.util.Random;

public class RNormTest
{    
    public static void main(String[] args)
    {
        int n = 10000000;
        testMethod(n, new Method() {
            void run() {
                boxMuller();
            }
            String name() {
                return "Box-Muller";
            }
        });
        
        testMethod(n, new Method() {
            void run() {
                boxMuller();
            }
            String name() {
                return "Polar rejection";
            }
        });
    }
    
    static class Pair
    {
        public Pair(double x, double y)
        {
            this.x = x;
            this.y = y;
        }
        public double x, y;
    }
    
    static abstract class Method
    {
        abstract void run();
        abstract String name();
    }

    
    static Pair boxMuller()
    {
        double U1 = r.nextDouble();
        double U2 = r.nextDouble();
        double root = sqrt( -2 * log(U1) );
        double N1 = root * cos(2 * PI * U2);
        double N2 = root * sin(2 * PI * U2);
        return new Pair(N1, N2);
    }
    
    static Pair polarRejection()
    {
        double V1 = 0, V2 = 0, W = 5;
        while (W > 1)
        {
            V1 = r.nextDouble();
            V2 = r.nextDouble();
            V1 = 2*V1 - 1;
            V2 = 2*V2 - 1;
            W = V1*V1 + V2*V2;
        }        
        double root = sqrt( -2 * log(W) / W);
        double N1 = root * V1;
        double N2 = root * V2;
        return new Pair(N1, N2);
    }
    
    static void testMethod(int n, Method m)
    {
        long start, end;
        start = System.currentTimeMillis();
        for (int i = 0; i < n; ++i)
            m.run();
        end = System.currentTimeMillis();
        System.out.println(m.name() + ": " +
                (double)(end-start)/1000 + " s.");
    }
    
    static Random r = new Random(1);
}
